File: 731. My Calendar II/solution.py
class MyCalendarTwo:

    def __init__(self):
        self.bookings = {}

    def book(self, start: int, end: int) -> bool:
        self.bookings[start] = self.bookings.get(start, 0) + 1
        self.bookings[end] = self.bookings.get(end, 0) - 1

        active_bookings = 0

       
        for time in sorted(self.bookings):
            active_bookings += self.bookings[time]
            if active_bookings >= 3:
                self.bookings[start] -= 1
                self.bookings[end] += 1
                return False

        return True

File: 731. My Calendar II/solution.js
var MyCalendarTwo = function() {
    this.bookings = [];
    this.overlaps = [];
};

/** 
 * @param {number} start 
 * @param {number} end
 * @return {boolean}
 */
MyCalendarTwo.prototype.book = function(start, end) {
    for (let [s, e] of this.overlaps) {
        if (start < e && end > s) {
            return false;
        }
    }

    for (let [s, e] of this.bookings) {
        if (start < e && end > s) {
            this.overlaps.push([Math.max(start, s), Math.min(end, e)]);
        }
    }

    this.bookings.push([start, end]);
    return true;
};
